drivers/pppou.c

```diff
@@ -23,109 +22,199 @@
-
-#if 0
-#define log_debug(fmt, ...)     do { if (1) pppos_printf(state, fmt, ##__VA_ARGS__); } while (0)
-#define log_at(fmt, ...)     	do { if (1) pppos_printf(state, COL_CYAN fmt COL_NORMAL, ##__VA_ARGS__); } while (0)
-#define log_info(fmt, ...)      do { if (1) pppos_printf(state, COL_CYAN fmt COL_NORMAL, ##__VA_ARGS__); } while (0)
-#define log_warn(fmt, ...)      do { if (1) pppos_printf(state, COL_YELLOW fmt COL_NORMAL, ##__VA_ARGS__); } while (0)
-#define log_error(fmt, ...)     do { if (1) pppos_printf(state, COL_RED  fmt COL_NORMAL, ##__VA_ARGS__); } while (0)
-
-static void pppos_printf(pppos_priv_t *state, const char *format, ...)
-{
-	char buf[256];
-	va_list arg;
+	unsigned char main_loop_stack[4096] __attribute__((aligned(8)));
+} pppou_priv_t;
 
-	va_start(arg, format);
-	vsnprintf(buf, sizeof(buf), format, arg);
-	va_end(arg);
 
-	printf("lwip: ppp@%s %s\n", state->serialdev_fn, buf);
-}
-#else
+#ifdef LOG_ENABLED
 
 #define log_debug(fmt, ...) syslog(LOG_DEBUG, fmt, ##__VA_ARGS__)
 #define log_at(fmt, ...)    syslog(LOG_INFO, fmt, ##__VA_ARGS__)
 #define log_info(fmt, ...)  syslog(LOG_INFO, fmt, ##__VA_ARGS__)
 #define log_warn(fmt, ...)  syslog(LOG_WARNING, fmt, ##__VA_ARGS__)
 #define log_error(fmt, ...) syslog(LOG_ERR, fmt, ##__VA_ARGS__)
 
+#else
+
+#define log_debug(fmt, ...)
+#define log_at(fmt, ...)
+#define log_info(fmt, ...)
+#define log_warn(fmt, ...)
+#define log_error(fmt, ...)
+
 #endif
 
-#define PPPOS_READ_AT_TIMEOUT_STEP_MS 		5
-#define PPPOS_READ_DATA_TIMEOUT_STEP_MS 	10
+#define PPPOU_READ_DATA_TIMEOUT_STEP_MS  10
+
+#define PPPOU_TRYOPEN_SERIALDEV_SEC      3
+#define PPPOU_CONNECT_RETRY_SEC          5
 
-#define PPPOS_TRYOPEN_SERIALDEV_SEC 		3
-#define PPPOS_CONNECT_RETRY_SEC 		5
-#define PPPOS_CONNECT_CMD_RETRY_MS		500
 
 /****** serial handling ******/
 
-static void serial_close(int fd)
+
+struct serial_speed_s {
+	const char *str;
+	speed_t speed;
+};
+
+
+static int _speedcmp(const void *l, const void *r)
 {
-	log_info("close()");
+	return strcmp(((struct serial_speed_s *)l)->str, ((struct serial_speed_s *)r)->str);
+}
 
-	if (fd >= 0)
-		close(fd);
 
-//	state->fd = -1;
-	// NOTE: set DISCONNECTED in status callback
-	// state->conn_state = CONN_STATE_DISCONNECTED;
+static speed_t serial_speed_from_string(const char *str)
+{
+	/* table is lexically sorted by `str` */
+	static const struct serial_speed_s stab[] = {
+		{ "0", B0 },           { "110", B110 },       { "115200", B115200 },
+		{ "1200", B1200 },     { "134", B134 },       { "150", B150 },
+		{ "1800", B1800 },     { "19200", B19200 },   { "200", B200 },
+		{ "230400", B230400 }, { "2400", B2400 },     { "300", B300 },
+		{ "38400", B38400 },   { "460800", B460800 }, { "4800", B4800 },
+		{ "50", B50 },         { "57600", B57600 },   { "600", B600 },
+		{ "75", B75 },         { "9600", B9600 },
+	};
+
+	struct serial_speed_s *sptr;
+
+	if (str && (sptr = bsearch(&str, stab, sizeof(stab) / sizeof(stab[0]), sizeof(stab[0]), _speedcmp)))
```

Comment:
a little bit of over-engineering, simple `for` loop would be probably enough (or even a switch-case block) :sunglasses:

Reply:
Ack.